auth/src/main/java/com/firebase/ui/auth/ui/accountlink/WelcomeBackIdpPrompt.java

```diff
@@ -127,83 +123,71 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
-        if (newIdpResponse == null) {
+    public void onSuccess(final IdpResponse idpResponse) {
+        if (idpResponse == null) {
             return; // do nothing
         }
 
-        AuthCredential newCredential = AuthCredentialHelper.getAuthCredential(newIdpResponse);
+        AuthCredential newCredential = AuthCredentialHelper.getAuthCredential(idpResponse);
         if (newCredential == null) {
             Log.e(TAG, "No credential returned");
             finish(ResultCodes.CANCELED, IdpResponse.getErrorCodeIntent(ErrorCodes.UNKNOWN_ERROR));
             return;
         }
 
-        final FirebaseAuth firebaseAuth = mActivityHelper.getFirebaseAuth();
-        FirebaseUser currentUser = firebaseAuth.getCurrentUser();
-
+        FirebaseUser currentUser = mActivityHelper.getCurrentUser();
         if (currentUser == null) {
-            Task<AuthResult> authResultTask = firebaseAuth.signInWithCredential(newCredential);
-            authResultTask.addOnCompleteListener(new OnCompleteListener<AuthResult>() {
-                @Override
-                public void onComplete(@NonNull Task<AuthResult> task) {
-                    if (task.isSuccessful() && mPrevCredential != null) {
-                        FirebaseUser firebaseUser = task.getResult().getUser();
-                        firebaseUser.linkWithCredential(mPrevCredential);
-                        firebaseAuth.signOut();
-                        firebaseAuth
-                                .signInWithCredential(mPrevCredential)
-                                .addOnFailureListener(new TaskFailureLogger(
-                                        TAG, "Error signing in with previous credential"))
-                                .addOnCompleteListener(new FinishListener(newIdpResponse));
-                    } else {
-                        finish(ResultCodes.OK, IdpResponse.getIntent(newIdpResponse));
-                    }
-                }
-            }).addOnFailureListener(
-                    new TaskFailureLogger(TAG, "Error signing in with new credential"));
+            mActivityHelper.getFirebaseAuth()
+                    .signInWithCredential(newCredential)
+                    .addOnSuccessListener(new OnSuccessListener<AuthResult>() {
+                        @Override
+                        public void onSuccess(AuthResult result) {
+                            if (mPrevCredential != null) {
+                                result.getUser()
+                                        .linkWithCredential(mPrevCredential)
+                                        .addOnFailureListener(new TaskFailureLogger(
+                                                TAG, "Error signing in with previous credential"))
+                                        .addOnCompleteListener(new FinishListener(idpResponse));
+                            } else {
+                                finish(ResultCodes.OK, IdpResponse.getIntent(idpResponse));
```

Comment:
I believe this was the original intention: if we have a previous credential, link it, else, finish ok. Previously, if we either didn't have a previous credential or the task failed, we would finish ok.

Because we would finish ok if we didn't have a previous credential, this isn't a behavior change.